firewood/src/merkle/node.rs

```diff
@@ -459,6 +463,203 @@ impl Storable for Node {
+            }
+            EncodedNodeType::Branch { children, value } => {
+                let mut list = <[Encoded<Vec<u8>>; BranchNode::MAX_CHILDREN + 1]>::default();
+
+                for (i, c) in children.iter().enumerate() {
+                    if let Some(c) = c {
+                        if c.len() >= TRIE_HASH_LEN {
+                            let serialized_hash =
+                                Bincode::serialize(&Keccak256::digest(c).to_vec())
+                                    .map_err(|e| S::Error::custom(format!("bincode error: {e}")))?;
+                            list[i] = Encoded::Data(serialized_hash);
+                        } else {
+                            list[i] = Encoded::Raw(c.to_vec());
+                        }
+                    }
+                }
+                if let Some(Data(val)) = &value {
+                    let serialized_val = Bincode::serialize(val)
+                        .map_err(|e| S::Error::custom(format!("bincode error: {e}")))?;
+                    list[BranchNode::MAX_CHILDREN] = Encoded::Data(serialized_val);
+                }
+
+                let mut seq = serializer.serialize_seq(Some(list.len()))?;
+                for e in list {
+                    seq.serialize_element(&e)?;
+                }
+                seq.end()
+            }
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for EncodedNode<Bincode> {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        use serde::de::Error;
+
+        let items: Vec<Encoded<Vec<u8>>> = Deserialize::deserialize(deserializer)?;
+        let len = items.len();
+        match len {
+            LEAF_NODE_SIZE => {
+                let mut items = items.into_iter();
+                let Some(Encoded::Raw(path)) = items.next() else {
+                    return Err(D::Error::custom(
+                        "incorrect encoded type for leaf node path",
+                    ));
+                };
+                let Some(Encoded::Raw(data)) = items.next() else {
+                    return Err(D::Error::custom(
+                        "incorrect encoded type for leaf node data",
+                    ));
+                };
+                let path_nibbles = Nibbles::<0>::new(&path);
+                let path = PartialPath::from_nibbles(path_nibbles.into_iter()).0;
+                let node = EncodedNodeType::Leaf(LeafNode {
+                    path,
+                    data: Data(data),
+                });
+                Ok(Self {
+                    node,
+                    phantom: PhantomData,
+                })
+            }
+            BranchNode::MSIZE => {
+                let mut children: [Option<Vec<u8>>; BranchNode::MAX_CHILDREN] = Default::default();
+                let mut value: Option<Data> = Default::default();
+                let len = items.len();
+
+                for (i, chd) in items.into_iter().enumerate() {
```

Comment:
nit: there are a few ways of handling the last item on  vector differently besides enumerating:

 - `let last = pop(items);` (although this requires a mutable items)
 - create an iterator from `items.chunks(MAX_CHILDREN)`, then the first item are the regular children and the last item is in a separate slice. The slices can form new iterators.

This is only a nit but it avoids the `mut children` and `mut value` as you can do map/collect for them.